### PR TITLE
feat: clone instead of fixed proxy

### PR DESCRIPTION
**Pull request by Foivos**
This saves 65k gas per interchainToken deployment and 100 gas per call to the interchain token. I deem it worthwile

### CODE REVIEW

File: contracts/utils/InterchainTokenDeployer.sol [hunk 1]
```diff
@@ -43,12 +42,18 @@ contract InterchainTokenDeployer is IInterchainTokenDeployer, Create3 {
         string calldata symbol,
         uint8 decimals
     ) external payable returns (address tokenAddress) {
-        bytes memory params = abi.encode(tokenManager, distributor, name, symbol, decimals);
         // slither-disable-next-line too-many-digits
-        bytes memory bytecode = bytes.concat(type(InterchainTokenProxy).creationCode, abi.encode(implementationAddress, params));
-
+        bytes memory bytecode = new bytes(0x37); //bytes.concat(type(InterchainTokenProxy).creationCode, abi.encode(implementationAddress, params));
```
Comment (milapsheth) [inline contracts/utils/InterchainTokenDeployer.sol:46]: add comments explaining the process

File: contracts/interchain-token/InterchainToken.sol [hunk 2]
```diff
@@ -17,21 +15,38 @@ import { Distributable } from '../utils/Distributable.sol';
...
+contract InterchainToken is BaseInterchainToken, ERC20Permit, Distributable, IInterchainToken {
     using AddressBytes for bytes;
 
     string public name;
     string public symbol;
     uint8 public decimals;
     address internal tokenManager_;
 
-    bytes32 private constant CONTRACT_ID = keccak256('interchain-token');
+    // bytes32(uint256(keccak256('interchain-token-is-setup-slot')) - 1);
```
Comment (milapsheth) [inline contracts/interchain-token/InterchainToken.sol:26]: Can we keep the `implementation()` method? Some explorers might not detect the minimal proxy

Comment (milapsheth) [inline contracts/interchain-token/InterchainToken.sol:26]:
```suggestion
    // bytes32(uint256(keccak256('interchain-token-initialized')) - 1);
```

File: contracts/interchain-token/InterchainToken.sol [hunk 3]
```diff
@@ -17,21 +15,38 @@ import { Distributable } from '../utils/Distributable.sol';
...
+
+    constructor() {
+        // Make the implementation act as if it has been setup already to disallow calls to init() (even though that wouldn't achieve anything really)
+        _setSetup();
+    }
+
+    /**
+     * @notice returns true if the contract has be setup.
+     */
+    function _isSetup() internal view returns (bool isSetup) {
```
Comment (milapsheth) [inline contracts/interchain-token/InterchainToken.sol:37]:
```suggestion
    function _isInitialized() internal view returns (bool initialized) {
```

File: contracts/interchain-token/InterchainToken.sol [hunk 4]
```diff
@@ -17,21 +15,38 @@ import { Distributable } from '../utils/Distributable.sol';
...
     using AddressBytes for bytes;
 
     string public name;
     string public symbol;
     uint8 public decimals;
     address internal tokenManager_;
 
-    bytes32 private constant CONTRACT_ID = keccak256('interchain-token');
+    // bytes32(uint256(keccak256('interchain-token-is-setup-slot')) - 1);
+    bytes32 internal constant IS_SETUP_SLOT = 0xb39f35de0a5b2620db9237c1e18c03b5e68a71236c9bdfbcd69f3582bab06df6;
```
Comment (milapsheth) [inline contracts/interchain-token/InterchainToken.sol:27]:
```suggestion
    bytes32 internal constant INITIALIZED_SLOT = 0xb39f35de0a5b2620db9237c1e18c03b5e68a71236c9bdfbcd69f3582bab06df6;
```

File: contracts/interchain-token/InterchainToken.sol [hunk 5]
```diff
@@ -17,21 +15,38 @@ import { Distributable } from '../utils/Distributable.sol';
...
+        }
+    }
 
     /**
-     * @notice Getter for the contract id.
+     * @notice sets the isSetup to true, to allow only a single setup.
      */
-    function contractId() external pure override returns (bytes32) {
-        return CONTRACT_ID;
+    function _setSetup() internal {
```
Comment (milapsheth) [inline contracts/interchain-token/InterchainToken.sol:46]:
```suggestion
    function _initialize() internal {
```

File: contracts/interchain-token/InterchainToken.sol [hunk 6]
```diff
@@ -44,20 +59,29 @@ contract InterchainToken is BaseInterchainToken, ERC20Permit, Implementation, Di
...
-        (tokenManagerAddress, distributor, tokenName, symbol, decimals) = abi.decode(params, (address, address, string, string, uint8));
+    function init(
+        address tokenManagerAddress,
+        address distributor,
+        string calldata tokenName,
+        string calldata tokenSymbol,
+        uint8 tokenDecimals
+    ) external {
+        if (_isSetup()) revert AlreadySetup();
+        _setSetup();
```
Comment (milapsheth) [inline contracts/interchain-token/InterchainToken.sol:76]:
```suggestion
        if (_isSetup()) revert AlreadySetup();

        _setSetup();
```

File: contracts/interfaces/IInterchainToken.sol [hunk 7]
```diff
@@ -12,14 +10,31 @@ import { IERC20Named } from './IERC20Named.sol';
 /**
  * @title IInterchainToken
  */
-interface IInterchainToken is IInterchainTokenStandard, IDistributable, IERC20MintableBurnable, IERC20Named, IImplementation {
+interface IInterchainToken is IInterchainTokenStandard, IDistributable, IERC20MintableBurnable, IERC20Named {
     error TokenManagerAddressZero();
     error TokenNameEmpty();
+    error AlreadySetup();
```
Comment (milapsheth) [inline contracts/interfaces/IInterchainToken.sol:16]:
```suggestion
    error AlreadyInitialized();
```

File: contracts/utils/InterchainTokenDeployer.sol (hunk 1 above)
Comment (milapsheth) [inline contracts/utils/InterchainTokenDeployer.sol:46]:
```suggestion
        bytes memory bytecode = new bytes(0x37);
```

File: contracts/utils/InterchainTokenDeployer.sol [hunk 8]
```diff
@@ -43,12 +42,18 @@ contract InterchainTokenDeployer is IInterchainTokenDeployer, Create3 {
...
+        address implementation = implementationAddress;
+        assembly {
+            mstore(add(bytecode, 0x20), shl(0x60, 0x3d602d80600a3d3981f3363d3d373d3d3d363d73))
```
Comment (milapsheth) [inline contracts/utils/InterchainTokenDeployer.sol:49]: can you create constants for the magic numbers here?

File: contracts/interchain-token/InterchainToken.sol (hunk 2 above)
Comment (Foivos) [reply contracts/interchain-token/InterchainToken.sol:26]: no we cannot

File: contracts/utils/InterchainTokenDeployer.sol (hunk 8 above)
Comment (Foivos) [reply contracts/utils/InterchainTokenDeployer.sol:52]: i don't see the point but we could

**Review by milapsheth — APPROVED** [review]

**Comment by re1ro** [comment]
Looks good and it's a huge improvement. Let's refactor CloneProxy into a lib.
Also would make sense to add comments for the bytecode constants from [ERC-1167](https://eips.ethereum.org/EIPS/eip-1167) and reference to the standard.

**Pull request merged**